### PR TITLE
[JENKINS-21017] When unmarshalling into an existing object, reset missing fields

### PR DESCRIPTION
**Pull request by jglick**
See [JENKINS-21017](https://issues.jenkins-ci.org/browse/JENKINS-21017). Rewrites #2736 with a somewhat different approach (but keeps its unit tests intact).

- [X] fix
- [X] unit tests
- [X] functional tests POSTing to `config.xml`

### Proposed changelog entries

* Updating Jenkins model objects (such as projects) by XML left fields at their old value if not defined in the new XML.

### Desired reviewers

@reviewbybees @jenkinsci/code-reviewers

### CODE REVIEW

File: test/src/test/java/hudson/model/ViewTest.java [hunk 1]
```diff
@@ -243,6 +243,24 @@
...
+        String xml = wc.goToXml("view/v/config.xml").getWebResponse().getContentAsString();
+        assertThat(xml, containsString("<description>one</description>"));
+        xml = xml.replace("<description>one</description>", "");
+        WebRequest req = new WebRequest(wc.createCrumbedUrl("view/v/config.xml"), HttpMethod.POST);
+        req.setRequestBody(xml);
+        req.setEncodingType(null);
+        wc.getPage(req);
+        assertEquals(null, view.getDescription()); // did not work
+        xml = new XmlFile(Jenkins.XSTREAM, new File(j.jenkins.getRootDir(), "config.xml")).asString();
+        assertThat(xml, not(containsString("<description>"))); // did not work
```
Comment (jglick) [inline test/src/test/java/hudson/model/ViewTest.java]: Note that `View` and `AbstractItem` have rather different implementations. `View` tries to load in memory, then `save`s to disk on success; `AbstractItem` saves to a temporary file, then loads from it, then renames to the real file on success. Not sure if there is a reason for the discrepancy, but anyway I am testing the effect on both implementations.

**Comment by jglick** [comment]
Lots of test failures like

```
java.lang.NullPointerException
	at jenkins.model.Jenkins$14.run(Jenkins.java:3057)
	at …
```

that I need to investigate.

**Comment by jglick** [comment]
Turns out that `Jenkins` has a ton of `final` fields with initializers, so if `loadConfig` nulls out undefined fields from old `config.xml`s, we get problems. These could be fixed one by one, but it seems a potentially more general compatibility issue. I will instead look into making the new behavior opt-in, to be used only by code which deliberately unmarshals in-place objects from user-supplied XML: `AbstractItem` and `View`. (`Computer` unmarshals user-supplied XML, but always creates fresh objects, so it is unaffected by JENKINS-21017.)

**Review by abayer — APPROVED** [review]

**Review by ghost — APPROVED** [review]
:bee: thanks for this one, Jesse!  I'm sure you had loads of fun with it ;)

**Review by stephenc — APPROVED** [review]
Pity it has to be opt-in... going to be a raft of down-stream PRs one by one

**Comment by jglick** [comment]
> going to be a raft of down-stream PRs one by one

Well, not that I can foresee. I know of one proprietary plugin which can benefit from using the new API.

**Comment by jglick** [comment]
@reviewbybees done

**Comment by daniel-beck** [comment]
KK is currently releasing 2.97, let's hold this until that's out just in case.

**Comment by oleg-nenashev** [comment]
`on-hold` till Jan 02 as discussed with @jglick

**Comment by jglick** [comment]
> till Jan 02

Well, it could be merged next week for the Dec 31/Jan 01 release (if one is made).

**Comment by oleg-nenashev** [comment]
Agreed, will merge it next Tuesday then

**Comment by daniel-beck** [comment]
@daspilker FYI this goes into 2.99, in case job-dsl docs need updating.

**Pull request merged**